src/main/java/com/palantir/common/streams/ImmutableCollectors.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2016 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.common.streams;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collector;
+
+/**
+ * Java 8 {@link Collector} for Guava immutable collections.
+ */
+public class ImmutableCollectors {
+
+    /*
+     * This collector has similar semantics to {@link Collectors::toSet} except that the set will be an {@link
+     * ImmutableSet}.
+     */
+    public static <T> Collector<T, ?, Set<T>> toSet() {
+        return Collector.of(
+                ImmutableSet::<T>builder,
+                ImmutableSet.Builder::<T>add,
+                (left, right) -> left.addAll(right.build()),
+                ImmutableSet.Builder::build,
+                Collector.Characteristics.UNORDERED);
+    }
+
+    /*
+     * This collector has similar semantics to {@link Collectors::toList} except that the list will be an {@link
+     * ImmutableList}.
+     */
+    public static <T> Collector<T, ?, ImmutableList<T>> toList() {
+        return Collector.of(
+                ImmutableList::<T>builder,
+                ImmutableList.Builder::<T>add,
+                (left, right) -> left.addAll(right.build()),
+                ImmutableList.Builder::build);
+    }
+
+    /*
+     * This collector has similar semantics to {@link Collectors::toMap} except that the map will be an {@link
+     * ImmutableMap}. Duplicate keys will cause an error.
+     */
+    public static <T, K, V> Collector<T, ?, Map<K, V>> toMap(Function<T, K> keyFunction, Function<T, V> valueFunction) {
```

Comment:
A fairly common way to use `Collectors` is with static imports; where instead of

    mySet.stream().filter(x -> x % 2 == 0).collect(Collectors.toSet());

you do

    mySet.stream().filter(x -> x % 2 == 0).collect(toSet());

for brevity and readability. I can point you to some internal code that does this, if you'd like (one of the projects I'm thinking about definitely uses palantir-streams). For people who do this, your methods add a little bit of confusion.

How much worse would it be for it to be

    ImmutableCollectors.toImmutableMap?

or maybe

    MoreCollectors.toImmutableMap ?